File: cakebot/Slots.py
"""
Cakebot - A fun and helpful Discord bot
Copyright (C) 2019-current year  Reece Dunham

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as published
by the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""

from random import choice, randint

emojis = "🍎🍊🍐🍋🍉🍇🍒"  #: The emojis that the game picks from.


def row():
    # type: () -> list
    """
    Returns a row of 3 emojis (as an array).

    Example: ["🍊", "🍉", "🍇"]
    """

    return [choice(emojis), choice(emojis), choice(emojis)]


def result():
    # type: () -> list
    """
    Returns the result of the game.

    This will be an array, consisting of the following data:
    - A number (either 0 or 1), 0 means you lose and 1 means you win
    - The row of emojis selected (see the row function documentation)
    """

    r = row()

    if r[0] == r[1] == r[2]:
        return [1, r]

    if randint(0, 5) == 4:
        # give player actual chance
        ic = str(choice(emojis))
        return [1, [ic, ic, ic]]

    return [0, r]

File: src/data/slots.ts
/**
 * Cakebot - A fun and helpful Discord bot
 * Copyright (C) 2021-current year  Reece Dunham
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import random from "random"

/**
 * All the possible emojis.
 */
export const emojis = ["🍎", "🍊", "🍐", "🍋", "🍉", "🍇", "🍒"]

/**
 * This is a special character. It is *not* a space, but instead
 * the braille character with no dots raised (effectively making
 * it bypass Discord's trimmer for spaces before the actual message
 * being sent.)
 */
export const BRAILLE_ZERO_DOT_CHAR = "⠀"

interface SlotRoll {
    row1: string
    row2: string
    row3: string
    winningRoll: boolean
    displayTip: boolean
}

/**
 * Get a random emoji.
 */
const getEmoji = () => emojis[random.int(0, emojis.length - 1)]

/**
 * Roll a slots game.
 */
export default function roll(): SlotRoll {
    // todo: allow this to be configured
    const WIN_CHANCE = 50

    const win: boolean = random.int(0, WIN_CHANCE) === 20

    return {
        winningRoll: win,
        displayTip: random.int(0, 20) === 15,
        row1: `${BRAILLE_ZERO_DOT_CHAR}${BRAILLE_ZERO_DOT_CHAR} ${getEmoji()}${getEmoji()}${getEmoji()}`,
        row2: `:arrow_forward: ${getEmoji()}${getEmoji()}${getEmoji()} :arrow_backward:`,
        row3: `${BRAILLE_ZERO_DOT_CHAR}${BRAILLE_ZERO_DOT_CHAR} ${getEmoji()}${getEmoji()}${getEmoji()}\n\n${
            win ? "**You win!**" : "You Lose. Better luck next time."
        }`,
    }
}
